### ISSUE TITLE
Getting error when trying to transcode from Explicit Big Endian to JPEGLossLess transfer syntax

### ISSUE THREAD

**Issue by quintonn** [dcm4che/dcm4che#868]
I have created a utility that will transcode/compress dicom messages from one transfer syntax to another.  
I've got it working for a number of use cases that our test team provided.  
But today I tried to add a number of additional edge cases and most of them fail with the same error.  
The error I am seeing is the following:  
> ERROR: 16-bit data must be encoded with syntax 1.2.840.10008.1.2.4.57 or 1.2.840.10008.1.2.4.70  

All the tests I added involve transcoding from either Big Endian or one of the Little Endian transfer syntaxes to a variety of the JPEG syntaxes.

To focus on 1 specific case for this ticket, I am converting from UID.ExplicitVRBigEndian to UID.JPEGLossless, which produces the above error.  

I am using very similar code to what is in the [dcm2dcm tool](https://github.com/dcm4che/dcm4che/blob/de21df71c491d807f7f2377dc6d56d19e9da3627/dcm4che-tool/dcm4che-tool-dcm2dcm/src/main/java/org/dcm4che3/tool/dcm2dcm/Dcm2Dcm.java#L403)

**To Reproduce**
To reproduce this error, attempt to transcode a dicom file that is in explicit big endian to JPEGLossless.

Other transfer syntaxes I attempted to convert to but failed with the same error are:  
- JPEGExtended12Bit
- JPEGBaseline8Bit
- JPEGLosslessSV1

And this failed for ImplicitVRLittleEndian, ExplicitVRBigEndian and ImplicitVRLittleEndian.  

The stacktrace that I am getting is:
> javax.imageio.IIOException: Native JPEG encoding error
	at org.dcm4che3.opencv.NativeJPEGImageWriter.write(NativeJPEGImageWriter.java:167)
	at org.dcm4che3.imageio.codec.Transcoder.compressFrame(Transcoder.java:688)
	at org.dcm4che3.imageio.codec.Transcoder.compressPixelData(Transcoder.java:530)
	at org.dcm4che3.imageio.codec.Transcoder.processPixelData(Transcoder.java:456)
	at org.dcm4che3.imageio.codec.Transcoder.access$600(Transcoder.java:71)
	at org.dcm4che3.imageio.codec.Transcoder$1.readValue(Transcoder.java:413)
	at org.dcm4che3.io.DicomInputStream.readAttributes(DicomInputStream.java:562)
	at org.dcm4che3.io.DicomInputStream.readAllAttributes(DicomInputStream.java:469)
	at org.dcm4che3.imageio.codec.Transcoder.transcode(Transcoder.java:388)
	.. my code ..
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: CvException [org.opencv.core.CvException: cv::Exception: OpenCV(4.0.0-dev) C:\jenkins\workspace\opencv\modules\imgcodecs\src\loadsave.cpp:1205: error: (-215:Assertion failed) code in function 'cv::dicomJpgWrite'
]
	at org.opencv.imgcodecs.Imgcodecs.dicomJpgWrite_0(Native Method)
	at org.opencv.imgcodecs.Imgcodecs.dicomJpgWrite(Imgcodecs.java:260)
	at org.dcm4che3.opencv.NativeJPEGImageWriter.write(NativeJPEGImageWriter.java:153)
	... 34 more

I am no expert on the dicom spec and it occurred to me that maybe some transfer syntaxes cannot be converted to others, as in the case of a compressed syntax into an uncompressed syntax.  
Maybe that is what is happening here too?

**Comment by gunterze** [comment]
Can you compress the image to JPEG Loss Less by utility `dcm2dcm --jpll` ?

**Comment by quintonn** [comment]
How do I get the dcm2dcm utility to run on my computer?  

But I did manage to run it from Eclipse. I passed "--jpll" as the Program arguments as well as the input dcm file in EBE and an output destination file path.  
I also had to set the "-Djava.library.path=" value the VM arguments to get open_cv libraries to run.  

But when I run it in Eclipse I get what looks like the same error:
> Jan 15, 2021 9:40:47 AM org.opencv.osgi.OpenCVNativeLoader init
INFO: Successfully loaded OpenCV native library.
log4j:WARN No appenders could be found for logger (org.dcm4che3.imageio.codec.Transcoder).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
ERROR: 16-bit data must be encoded with syntax 1.2.840.10008.1.2.4.57 or 1.2.840.10008.1.2.4.70 
Failed to transcode C:\projects\dcm4che\ebe_message1.dcm: Native JPEG encoding error
javax.imageio.IIOException: Native JPEG encoding error
	at org.dcm4che3.opencv.NativeJPEGImageWriter.write(NativeJPEGImageWriter.java:167)
	at org.dcm4che3.imageio.codec.Transcoder.compressFrame(Transcoder.java:688)
	at org.dcm4che3.imageio.codec.Transcoder.compressPixelData(Transcoder.java:530)
	at org.dcm4che3.imageio.codec.Transcoder.processPixelData(Transcoder.java:456)
	at org.dcm4che3.imageio.codec.Transcoder.access$600(Transcoder.java:71)
	at org.dcm4che3.imageio.codec.Transcoder$1.readValue(Transcoder.java:413)
	at org.dcm4che3.io.DicomInputStream.readAttributes(DicomInputStream.java:562)
	at org.dcm4che3.io.DicomInputStream.readAllAttributes(DicomInputStream.java:469)
	at org.dcm4che3.imageio.codec.Transcoder.transcode(Transcoder.java:388)
	at org.dcm4che3.tool.dcm2dcm.Dcm2Dcm.transcodeWithTranscoder(Dcm2Dcm.java:410)
	at org.dcm4che3.tool.dcm2dcm.Dcm2Dcm.transcode(Dcm2Dcm.java:350)
	at org.dcm4che3.tool.dcm2dcm.Dcm2Dcm.mtranscode(Dcm2Dcm.java:341)
	at org.dcm4che3.tool.dcm2dcm.Dcm2Dcm.mtranscode(Dcm2Dcm.java:318)
	at org.dcm4che3.tool.dcm2dcm.Dcm2Dcm.main(Dcm2Dcm.java:290)
Caused by: CvException [org.opencv.core.CvException: cv::Exception: OpenCV(4.0.0-dev) C:\jenkins\workspace\opencv\modules\imgcodecs\src\loadsave.cpp:1205: error: (-215:Assertion failed) code in function 'cv::dicomJpgWrite'
]
	at org.opencv.imgcodecs.Imgcodecs.dicomJpgWrite_0(Native Method)
	at org.opencv.imgcodecs.Imgcodecs.dicomJpgWrite(Imgcodecs.java:260)
	at org.dcm4che3.opencv.NativeJPEGImageWriter.write(NativeJPEGImageWriter.java:153)
	... 13 more

But one of my colleagues seems to think that according to the Dicom spec it is not or should not be possible to compress Big Endian, but the wording is a bit confusing on whether or not it's not possible or not recommended.  

But even if that is the case, my code for converting ELE and ILE is also failing.  
I also checked and running the dcm2dcm code with --jpll on the ELE and ILE files create the same error.

**Comment by nroduit** [comment]
Currently, when a conversion is not consistent an exception is triggered. Here the problem is that the process try to convert an image with a 16-bit pixel depth to 8-bit which results in big loss of information (usually not desirable).

I proposed a [PR](https://github.com/dcm4che/dcm4che/pull/825) which give a way to do any conversion and if the transcoding is not consistent  then the syntax is change automatically. This way it is possible to convert a large number of images without any errors. The rules are:
- convert 1.2.840.10008.1.2.4.50 to 1.2.840.10008.1.2.4.70 when bitstored is more than 8
- convert 1.2.840.10008.1.2.4.51, 1.2.840.10008.1.2.4.53, 1.2.840.10008.1.2.4.55 to 1.2.840.10008.1.2.4.70 when bitstored is less than 13 and with signed data
- uncompress when bitstored is more than 16

**Comment by gunterze** [comment]
Cannot reproduce - neither before nor after applying #825:
```
curl https://www.dclunie.com/images/signedrange/signedrangeimages.tar.bz2 \
  | tar xj IMAGES/test16unsigned/test16unsigned.dcm
dcm2dcm -t 1.2.840.10008.1.2.2 IMAGES/test16unsigned/test16unsigned.dcm bigendian16unsigned.dcm
dcm2dcm --jpll bigendian16unsigned.dcm jpll16unsigned.dcm
Native OpenCV library has been found in /home/gunter/dcm4che/lib/linux-x86-64/libopencv_java.so
Jan 17, 2021 1:17:13 PM org.opencv.osgi.OpenCVNativeLoader init
INFO: Successfully loaded OpenCV native library.
13:17:13,553 DEBUG - Compressor: org.dcm4che3.opencv.NativeJPEGImageWriter
13:17:13,599 DEBUG - Compressed frame #1 in 18 ms, ratio 7.3040957:1
```

**Comment by nroduit** [comment]
Yes with `--jpll` all conversions are possible. According to the error above, I guess another jpeg syntax is used. @quintonn  I would recommend to use the dcm2dcm tool because this part of transcoding is quite tricky.

**Comment by quintonn** [comment]
@gunterze, you're not using the  `--jpll` flag?

**Comment by quintonn** [comment]
Where can I get a dcm2dcm exe to run on windows?
I want to make sure I am running the correct version of it?
The only way I can run it currently is from Eclipse by passing in the arguments using Eclipse, but I would like to run it from command line.

**Comment by hczedik** [comment]
Hi @quintonn,
the easiest way to run the dcm2dcm tool is, to download the "Binary distribution" of the newest dcm4che release (see https://github.com/dcm4che/dcm4che/releases).
Unzip the downloaded file, and, in the Windows command line, run the bin\dcm2dcm.bat file.

**Comment by quintonn** [comment]
@hczedik thanks.
I got it now.

**Comment by quintonn** [comment]
@gunterze using the dcm2dcm2 tool I am able to convert all my dicom files.  
But interestingly, running the dcm2dcm tool from Eclipse does not work.  
Also, I created a new unit test where I copied the exact code from this repository to transcode and it also doesn't work.  

I suspect it is related to a missing jar, or the 3rd party imaging libraries or perhaps it's related to being run from inside Eclipse.  

I will do a bit more testing to see what other differences I can find.

**Comment by quintonn** [comment]
@gunterze 
Yes, that seems to have been the problem.  
I noticed the **opencv_java.dll** file I had was significantly smaller than the one include in the zip file you provided above.  
I replaced mine with yours and many of the errors went away.  

What is the best way to make sure I always have to most up-to-date opencv library files?  
Is there a CI/CD way, (without Maven, we can't use that) ?

But there are still 2 transfer syntaxes I cannot transcode to and give the same exception (both in my unit tests and from the command line using the `-t` flag), and they are:  
- 1.2.840.10008.1.2.4.50
- 1.2.840.10008.1.2.4.51

Maybe these 2 can't be transcoded/compressed?
